client/build_test.go

```diff
@@ -718,6 +722,246 @@ func testClientGatewayContainerMounts(t *testing.T, sb integration.Sandbox) {
+		time.Sleep(100 * time.Millisecond)
+		inputW.Write([]byte("cat /tmp/newfile\n"))
+		time.Sleep(500 * time.Millisecond)
+		err = pid1.Resize(ctx, client.WinSize{
+			Rows: 60,
+			Cols: 100,
+		})
+		require.NoError(t, err)
+		time.Sleep(500 * time.Millisecond)
+
+		inputW.Write([]byte("ttysize\n"))
+		time.Sleep(100 * time.Millisecond)
+		inputW.Write([]byte("exit 99\n"))
+
+		err = pid1.Wait()
+		var exitError *errdefs.ExitError
+		require.True(t, errors.As(err, &exitError))
+		require.Equal(t, uint32(99), exitError.ExitCode)
+
+		return &client.Result{}, err
+	}
+
+	_, err = c.Build(ctx, SolveOpt{}, product, b, nil)
+	require.Error(t, err)
+
+	// filter out stray terminal CSR codes
+	badControl := regexp.MustCompile(`\r.*\x1b\[J`)
+	// trim out \r so we can cleanly split on newline
+	filteredOutput := strings.TrimSpace(strings.ReplaceAll(badControl.ReplaceAllString(output.String(), ""), "\r", ""))
+
+	t.Logf("OUTPUT: %s", filteredOutput)
+	outputLines := strings.Split(filteredOutput, "\n")
+
+	require.Equal(t, []string{
+		"% ttysize",
+		"80 40",
+		"% cd /tmp",
+		"% pwd",
+		"/tmp",
+		"% echo foobar > newfile",
+		"% cat /tmp/newfile",
+		"foobar",
+		"% ttysize",
+		"100 60",
+		"% exit 99",
+	}, outputLines)
+
+	inputW.Close()
+	inputR.Close()
+
+	checkAllReleasable(t, c, sb, true)
+}
+
+// testClientGatewayContainerExecTty is testing that we can get a tty via
+// executor.Exec (secondary process)
+func testClientGatewayContainerExecTty(t *testing.T, sb integration.Sandbox) {
+	if sb.Rootless() {
+		// TODO fix this
+		// We get `panic: cannot statfs cgroup root` when running this test
+		// with runc-rootless
+		t.Skip("Skipping runc-rootless for cgroup error")
+	}
+	requiresLinux(t)
+	ctx := context.TODO()
+
+	c, err := New(ctx, sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	product := "buildkit_test"
+
+	inputR, inputW := io.Pipe()
+	output := bytes.NewBuffer(nil)
+	b := func(ctx context.Context, c client.Client) (*client.Result, error) {
+		st := llb.Image("busybox:latest")
+
+		def, err := st.Marshal(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to marshal state")
+		}
+
+		r, err := c.Solve(ctx, client.SolveRequest{
+			Definition: def.ToPB(),
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to solve")
+		}
+
+		ctr, err := c.NewContainer(ctx, client.NewContainerRequest{
+			Mounts: []client.Mount{{
+				Dest:      "/",
+				MountType: pb.MountType_BIND,
+				Ref:       r.Ref,
+			}},
+		})
+		require.NoError(t, err)
+
+		pid1, err := ctr.Start(ctx, client.StartRequest{
+			Args: []string{"sleep", "10"},
+		})
+		require.NoError(t, err)
+
+		defer pid1.Wait()
+		defer ctr.Release(ctx)
+
+		pid2, err := ctr.Start(ctx, client.StartRequest{
+			Args:   []string{"sh"},
+			Tty:    true,
+			Stdin:  inputR,
+			Stdout: &nopCloser{output},
+			Stderr: &nopCloser{output},
+			Env:    []string{"PS1=% "},
+		})
+		require.NoError(t, err)
+
+		err = pid2.Resize(ctx, client.WinSize{
```

Comment:
most of this test looks very similar. can we put the shared part in a helper function or use subtest.

Reply:
updated to add common routine for the terminal input and output